### PR TITLE
[TASK] Rewrite questionnaires in Vue

### PR DESCRIPTION
**Pull request by Mark-de-Haan**
#### Checklist
- [ ] Functionality works & meets specifications
- [ ] Code reviewed
- [ ] Code unit/integration/system tested
- [ ] User documentation updated
- [ ] (If you have changed REST API interface) view-swagger.ftl updated
- [ ] Test plan template updated
- [ ] Clean commits
- [ ] Added Feature/Fix to release notes
- [ ] Integration tests run correctly

### CODE REVIEW

**Comment by Mark-de-Haan** [comment]
# Review comments:
## Bugs
* Jumping to chapter via navigation triggers form validation
* Smaller window size breaks things
* On value change is triggered on initial load

## Style comments
* Loading indicator in Questionnaire list
* QuestionnaireStart component start back buttons a bit weird
* Start questionnaire text should be equal to status e.g. OPEN == continue
* Next button on top and bottom looks weird on smaller chapters
* Next in bigger chapters with something wrong is weird, why am I not going to next??
* Loading indicator in Next button
* Add questionnaire styling to error field messages
* Progress bar should be on progress, not on chapter number

## Nice to have:
* Loading indicator for Questionnaire takes long, maybe a bit more animated

File: molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/controller/QuestionnaireController.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+import static org.springframework.http.HttpStatus.OK;
+
+@Controller
+@RequestMapping(QuestionnaireController.URI)
+public class QuestionnaireController extends VuePluginController
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/controller/QuestionnaireController.java]: Add javadoc and or swagger docs to controller end points, these serve as a 'public' ( for app builder) api and should be documented.

File: molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/QuestionnaireService.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+import org.molgenis.questionnaires.response.QuestionnaireResponse;
+
+import java.util.List;
+
+public interface QuestionnaireService
+{
+	/**
+	 * Return a list of all questionnaires
+	 * Creates a questionnaire entry for the current user if it does not yet exist
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/QuestionnaireService.java]: i do not understand this line , why would a get/list result in a create ?

File: molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/QuestionnaireService.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	 * @return A List of {@link QuestionnaireResponse}
+	 */
+	List<QuestionnaireResponse> getQuestionnaires();
+
+	/**
+	 * Start a questionnaire based on ID
+	 * Sets the questionnaire status to 'OPEN'
+	 *
+	 * @param id The ID of a questionnaire
+	 * @return A {@link QuestionnaireResponse}
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/QuestionnaireService.java]: method says void , but java doc talks about returned value , i would suggest returning a value

File: molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+						  .eq(EntityTypeMetadata.EXTENDS, QUESTIONNAIRE)
+						  .findAll()
+						  .filter(entityType -> userPermissionEvaluator.hasPermission(
+								  new EntityTypeIdentity(entityType.getId()), EntityTypePermission.WRITE))
+						  .map(entityType ->
+						  {
+							  String entityTypeId = entityType.getId();
+							  Questionnaire questionnaireEntity = findQuestionnaireEntity(entityTypeId);
+
+							  if (questionnaireEntity == null)
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java]: Why is this needed , this seems like a side effect , i would suggest splitting it of to a separate method/service if needed at all.

File: molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+							  }
+							  return QuestionnaireResponse.create(questionnaireEntity);
+						  })
+						  .collect(toList());
+	}
+
+	@Override
+	public void startQuestionnaire(String id)
+	{
+		Questionnaire questionnaire = findQuestionnaireEntity(id);
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java]: what happens if no questionnaire is found for the give id

File: molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+		if (questionnaire.getStatus().equals(NOT_STARTED))
+		{
+			// Set questionnaire status to open once it has been requested
+			questionnaire.setStatus(OPEN);
+			dataService.update(id, questionnaire);
+		}
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java]: what about the else condition ?

File: molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	@Override
+	public void startQuestionnaire(String id)
+	{
+		Questionnaire questionnaire = findQuestionnaireEntity(id);
+		if (questionnaire.getStatus().equals(NOT_STARTED))
+		{
+			// Set questionnaire status to open once it has been requested
+			questionnaire.setStatus(OPEN);
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java]: the more oo approach would be to have a method on the questionnaire object had opens (of closes) the questionnaire, this keep the state internal (saver/ more self describing).

File: molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+		String key = id + "_submissionText";
+		String submissionText = staticContentService.getContent(key);
+
+		if (submissionText == null)
+		{
+			submissionText = DEFAULT_SUBMISSION_TEXT;
+			staticContentService.submitContent(key, submissionText);
+		}
+
+		return submissionText;
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java]: if submissionText is null you are returning null. this is a really bad idea, a it pushed the problem to the user of this method ( who does not know about this)

Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java]: maybe use strong type or docs to signal to user that html may be returned

File: molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+	/**
+	 * Find 1 row in the Questionnaire table that belongs to the current user
+	 *
+	 * @param entityTypeId The ID of a questionnaire table
+	 * @return An {@link Entity} of type {@link Questionnaire}
+	 */
+	private Questionnaire findQuestionnaireEntity(String entityTypeId)
+	{
+		return questionnaireFactory.create(dataService.findOne(entityTypeId, EQ(OWNER_USERNAME, getCurrentUsername())));
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java]: what is no Questionnaire is found for the given id ?

File: molgenis-questionnaires/src/main/frontend/src/store/index.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
+import Vue from 'vue'
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/frontend/src/store/index.js]: i would suggest not using index.js in molgenis , it makes is hard to find files as there are a lot of index.js files. This idea ( i do not like it in general) only really works for smaller ( modules) project imo.

File: molgenis-questionnaires/src/main/frontend/src/store/index.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+import Vuex from 'vuex'
+
+import actions from './actions'
+import getters from './getters'
+import mutations from './mutations'
+
+const state = {
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/frontend/src/store/index.js]: use flow typing for the store, makes it easier to reason about

File: molgenis-questionnaires/src/main/frontend/src/store/mutations.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
+const mutations = {
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/frontend/src/store/mutations.js]: all flow types to all method params

File: molgenis-questionnaires/src/main/frontend/src/store/getters.js [hunk 13]
```diff
@@ -0,0 +1,42 @@
+const getAllFields = (chapter, data) => {
+  return chapter.children.reduce((accumulator, child) => {
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/frontend/src/store/getters.js]: what if chapter has no children ? , adding flow typing would fix this

File: molgenis-questionnaires/src/main/frontend/src/store/getters.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    if (child.type === 'field-group') {
+      return [...accumulator, ...getAllFields(child, data)]
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/frontend/src/store/getters.js]: mind blown :)

File: molgenis-questionnaires/src/main/frontend/src/store/getters.js [hunk 15]
```diff
@@ -0,0 +1,42 @@
+const getAllFields = (chapter, data) => {
+  return chapter.children.reduce((accumulator, child) => {
+    if (child.type === 'field-group') {
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/frontend/src/store/getters.js]: we need more flow typing

File: molgenis-questionnaires/src/main/frontend/src/store/actions.js [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+
+      commit('SET_CHAPTER_FIELDS', chapters)
+      commit('SET_QUESTIONNAIRE_ROW_ID', data[response.meta.idAttribute])
+      commit('SET_FORM_DATA', form.formData)
+    })
+  },
+
+  /* eslint-disable no-empty-pattern */
+  'GET_QUESTIONNAIRE_OVERVIEW' ({}, questionnaireId) {
+    return api.get('/api/v2/' + questionnaireId)
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/frontend/src/store/actions.js]: why not use the store ?

File: molgenis-questionnaires/src/main/frontend/src/store/actions.js [hunk 17]
```diff
@@ -0,0 +1,77 @@
+import api from '@molgenis/molgenis-api-client'
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/frontend/src/store/actions.js]: need error handler for all actions

File: molgenis-questionnaires/src/main/frontend/src/components/QuestionnaireStart.vue [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+      questionnaireDescription () {
+        return this.$store.state.questionnaireDescription
+      },
+      questionnaireLabel () {
+        return this.$store.state.questionnaireLabel
+      }
+    },
+    created () {
+      this.$store.dispatch('START_QUESTIONNAIRE', this.questionnaireId).then(() => {
+        if (!this.$store.state.mapperOptions.booleanLabels) {
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/frontend/src/components/QuestionnaireStart.vue]: maybe move branches to separate methods , easer to read and test

File: molgenis-questionnaires/src/main/frontend/src/components/QuestionnaireChapter.vue [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+
+      /**
+       * Forces validation if a user wants to go to the next chapter
+       * Triggers client side validation by setting status to 'SUBMITTED'
+       */
+      navigateToNextChapter () {
+        this.$store.commit('UPDATE_FORM_STATUS', 'SUBMITTED')
+        this.formState.$submitted = true
+
+        this.$nextTick(() => {
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/frontend/src/components/QuestionnaireChapter.vue]: add comment on why this is needed

File: molgenis-questionnaires/src/main/frontend/src/components/QuestionnaireChapter.vue [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+      },
+
+      /**
+       * Redirect to thank you page when submit is successful
+       */
+      submitQuestionnaire () {
+        this.$store.commit('UPDATE_FORM_STATUS', 'SUBMITTED')
+        this.formState.$submitted = true
+
+        this.$nextTick(() => {
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/frontend/src/components/QuestionnaireChapter.vue]: add comment on why this is needed

File: molgenis-questionnaires/src/main/frontend/src/components/QuestionnaireChapter.vue [hunk 21]
```diff
@@ -0,0 +1,318 @@
...
+    methods: {
+      onValueChanged (formData) {
+        this.$store.commit('SET_FORM_DATA', formData)
+      },
+
+      /**
+       * Forces validation if a user wants to go to the next chapter
+       * Triggers client side validation by setting status to 'SUBMITTED'
+       */
+      navigateToNextChapter () {
```
Comment (ConnorStroomberg) [inline molgenis-questionnaires/src/main/frontend/src/components/QuestionnaireChapter.vue]: maybe change method name to indicate it is only the request to navigate and validation is done first

**Comment by ConnorStroomberg** [comment]
When i am logged out and logback in i get

`Attribute 'id' of entity 'c6_questionnaire' is readonly` 
on
`http://localhost:8080/menu/plugins/questionnaires/c6_questionnaire/chapter/4`

**Comment by ConnorStroomberg** [comment]
Would be good to add a 'happy flow' e2e test or selenium e2e-with backend

File: molgenis-questionnaires/src/main/frontend/src/store/actions.js (hunk 16 above)
Comment (Mark-de-Haan) [reply molgenis-questionnaires/src/main/frontend/src/store/actions.js]: I don't understand the comment. Overview screen is a separate readonly view of one row in the questionnaire table. If linked via URL, the overview component will need to fetch data. It could even work outside of the store.

File: molgenis-questionnaires/src/main/frontend/src/store/getters.js (hunk 14 above)
Comment (Mark-de-Haan) [reply molgenis-questionnaires/src/main/frontend/src/store/getters.js]: recursive reducers ftw 🤓

File: molgenis-questionnaires/src/main/frontend/src/store/getters.js (hunk 13 above)
Comment (Mark-de-Haan) [reply molgenis-questionnaires/src/main/frontend/src/store/getters.js]: Then you have an empty array, and all the components are designed to work with an empty array as "base" state. But flow typing should be added yes

File: molgenis-questionnaires/src/main/frontend/src/store/actions.js (hunk 17 above)
Comment (Mark-de-Haan) [reply molgenis-questionnaires/src/main/frontend/src/store/actions.js]: done

File: molgenis-questionnaires/src/main/frontend/src/store/mutations.js (hunk 12 above)
Comment (Mark-de-Haan) [reply molgenis-questionnaires/src/main/frontend/src/store/mutations.js]: done

File: molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java (hunk 5 above)
Comment (Mark-de-Haan) [reply molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java]: If questionnaire is null, a new entry is created for the current user

File: molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java (hunk 9 above)
Comment (Mark-de-Haan) [reply molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/service/impl/QuestionnaireServiceImpl.java]: It will return null, which is used to determine if a new row should be created for the current user

File: molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/controller/QuestionnaireController.java (hunk 1 above)
Comment (Mark-de-Haan) [reply molgenis-questionnaires/src/main/java/org/molgenis/questionnaires/controller/QuestionnaireController.java]: done

**Comment by ConnorStroomberg** [comment]
When loading the questionnaire, the long wait time is mostly due to wait for the server response. If we want to improve this we need to profile the request handling.

**Comment by ConnorStroomberg** [comment]
When pressing the next chapter button the ui becomes unresponsive, the js heap space is maxed out en most of the time is spend in the vue watch code. we need to reduce the number of watches to the ui become responsive. As a work around i suggest we add a spinner to the button.

**Comment by ConnorStroomberg** [comment]
test this please
